src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.State.cs

```diff
@@ -0,0 +1,266 @@
+            public ImmutableHashSet<string> CompilationReferences { get; }
+            public ImmutableHashSet<string> CopyReferenceInputs { get; }
+
+            private State()
+            {
+                ItemTypes = ImmutableHashSet.Create(StringComparers.ItemTypes);
+                Items = ImmutableDictionary.Create<string, ImmutableHashSet<(string path, string? link, CopyToOutputDirectoryType copyType)>>(StringComparers.ItemTypes);
+                CustomInputs = ImmutableHashSet.Create(StringComparers.Paths);
+                CustomOutputs = ImmutableHashSet.Create(StringComparers.Paths);
+                BuiltOutputs = ImmutableHashSet.Create(StringComparers.Paths);
+                CopiedOutputFiles = ImmutableDictionary.Create<string, string>(StringComparers.Paths);
+                AnalyzerReferences = ImmutableHashSet.Create(StringComparers.Paths);
+                CompilationReferences = ImmutableHashSet.Create(StringComparers.Paths);
+                CopyReferenceInputs = ImmutableHashSet.Create(StringComparers.Paths);
+            }
+
+            private State(
+                string msBuildProjectFullPath,
+                string msBuildProjectDirectory,
+                string? markerFile,
+                string outputRelativeOrFullPath,
+                string newestImportInput,
+                IComparable lastVersionSeen,
+                bool isDisabled,
+                ImmutableHashSet<string> itemTypes,
+                ImmutableDictionary<string, ImmutableHashSet<(string, string?, CopyToOutputDirectoryType)>> items,
+                ImmutableHashSet<string> customInputs,
+                ImmutableHashSet<string> customOutputs,
+                ImmutableHashSet<string> builtOutputs,
+                ImmutableDictionary<string, string> copiedOutputFiles,
+                ImmutableHashSet<string> analyzerReferences,
+                ImmutableHashSet<string> compilationReferences,
+                ImmutableHashSet<string> copyReferenceInputs)
+            {
+                MSBuildProjectFullPath = msBuildProjectFullPath;
+                MSBuildProjectDirectory = msBuildProjectDirectory;
+                MarkerFile = markerFile;
+                OutputRelativeOrFullPath = outputRelativeOrFullPath;
+                NewestImportInput = newestImportInput;
+                LastVersionSeen = lastVersionSeen;
+                IsDisabled = isDisabled;
+                ItemTypes = itemTypes;
+                Items = items;
+                CustomInputs = customInputs;
+                CustomOutputs = customOutputs;
+                BuiltOutputs = builtOutputs;
+                CopiedOutputFiles = copiedOutputFiles;
+                AnalyzerReferences = analyzerReferences;
+                CompilationReferences = compilationReferences;
+                CopyReferenceInputs = copyReferenceInputs;
+            }
+
+            public static State CreateEmpty()
+            {
+                return new State();
```

Comment:
How many of these are created? Could just be a static readonly Empty instance.

Reply:
Not many. Only on creation and on unload. I'll make an `Empty` property though.